### DISCUSSION TITLE
When I/O-related callback functions are executed

### DISCUSSION THREAD

**Discussion by dSalieri** [nodejs/help#4626] [Q&A]
The node.js documentation is very awesome and that's why I'm here :)

There are two places where the node.js documentation talks about executing I/O callbacks:
- **pending callbacks:** _executes I/O callbacks_ deferred to the next loop iteration.
- **poll:** retrieve new I/O events; _execute I/O related callbacks_ (almost all with the exception of close callbacks, the ones scheduled by timers, and setImmediate()); node will block here when appropriate.

Previously, in the old version of the documentation it was like this:
- **I/O callbacks:** executes almost all callbacks with the exception of close callbacks, the ones scheduled by timers, and setImmediate().
- **poll:** retrieve new I/O events; node will block here when appropriate.

Can anyone explain what's going on? And if in reality I/O callbacks are executed in two phases, please provide real code examples that describe this pattern.

It would also be nice (for anyone who understands the node.js source code) to have an explanation indicating the key lines of the source code (where the I/O callbacks are called, in what queue and when they are put, etc.)

I hope for an answer.

**Comment by mateuspaulino** [comment]
Hi @dSalieri 

In the pending callbacks phase, Node.js executes I/O callbacks that were deferred to the next loop iteration. These are callbacks that couldn't be processed in the previous iteration and are thus handled later.

In the poll phase, Node.js retrieves new I/O events and executes their callbacks, excluding close callbacks, those scheduled by timers, and setImmediate(). This is where most I/O operations' callbacks are executed. Node.js will also block and wait for new I/O events in this phase if necessary.

The previous documentation combined these explanations, but the new documentation clearly separates them for better understanding.

**Reply by dSalieri** [reply]
@mateuspaulino hello.

This is all interesting, of course, your answer is almost no different from the documentation.

I also asked for real examples with code (and of course they don't exist). Can your answer be considered an answer, I don't think so, sorry.

---

None of the node.js documentation explained the phases clearly, except for _timer_, _check_. The rest of the phases are just guess what API is used for them, there are not even descriptive groups to understand. And I am not even talking about the difference between **pending callbacks** and **poll**. There is not a single article on the network with code examples where you can clearly distinguish these stages (and I have not touched on such phases as _idle_ and _prepare_, which exist "just because", without any specific description of what they do)
